Fix signed integer overflow for nth_value() window function

azat opened the pull request:
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix signed integer overflow for nth_value() window function

Detailed description / Documentation draft:
CI report [1]:

    ../src/Processors/Transforms/WindowTransform.cpp:1699:54: runtime error: signed integer overflow: -9223372036854775808 - 1 cannot be represented in type 'long' Received signal -3 Received signal Unknown signal (-3)

In query:

```sql

    SELECT
        number,
        nth_value(number, 2147483648) OVER w,
        anyOrNull(number) OVER (ORDER BY -2147483648 - intDiv(number, 1024) DESC NULLS FIRST, number DESC ROWS BETWEEN 65535 FOLLOWING AND UNBOUNDED FOLLOWING),
        nth_value(number, 65537) OVER w AS firstValue,
        nth_value(number, -9223372036854775808) OVER w AS secondValue,
        nth_value(number, 1048576) OVER w AS thirdValue
    FROM numbers(1)
    WINDOW w AS (ORDER BY number DESC)
    ORDER BY number DESC

```

  [1]: https://clickhouse-test-reports.s3.yandex.net/28532/7623af5513e12aa8dfa1bee963caffe00185c31a/fuzzer_ubsan/report.html#fail1

Cc: @akuzm 
Fixes: #28001

azat committed:
Fix signed integer overflow for nth_value() window function

CI report [1]:

    ../src/Processors/Transforms/WindowTransform.cpp:1699:54: runtime error: signed integer overflow: -9223372036854775808 - 1 cannot be represented in type 'long' Received signal -3 Received signal Unknown signal (-3)

In query:

    SELECT
        number,
        nth_value(number, 2147483648) OVER w,
        anyOrNull(number) OVER (ORDER BY -2147483648 - intDiv(number, 1024) DESC NULLS FIRST, number DESC ROWS BETWEEN 65535 FOLLOWING AND UNBOUNDED FOLLOWING),
        nth_value(number, 65537) OVER w AS firstValue,
        nth_value(number, -9223372036854775808) OVER w AS secondValue,
        nth_value(number, 1048576) OVER w AS thirdValue
    FROM numbers(1)
    WINDOW w AS (ORDER BY number DESC)
    ORDER BY number DESC

  [1]: https://clickhouse-test-reports.s3.yandex.net/28532/7623af5513e12aa8dfa1bee963caffe00185c31a/fuzzer_ubsan/report.html#fail1

alexey-milovidov commented on src/Processors/Transforms/WindowTransform.cpp:
```diff
@@ -1696,22 +1696,16 @@ struct WindowFunctionNthValue final : public WindowFunction
+            transform->current_row.row].get<Int64>();
 
-        if (offset < 0)
+        if (offset > INT_MAX || offset <= 0)
```
Why INT_MAX while we have int64?

alexey-milovidov replied:
We still can have more than 2bn rows in a block.

azat replied:
Yes, initially this is not my code, but I guess it was done as some sort of soft limit.
Maybe @akuzm or @ryzuo can comment on this?
In the mean time I will increase it.

azat replied:
Also lagInFrame/leadInFrame has the same limits, after d9ad06bd6d3

azat committed:
Allow values up to INT64_MAX for nth_value/lagInFrame/leadInFrame

azat committed:
Use the same error for leadInFrame()/lagInFrame() as in nth_value()

azat committed:
Eliminate int64_t > INT64_MAX checks (found by PVS)

https://clickhouse-test-reports.s3.yandex.net/28773/b38311538979c0ffe27f6c8c5732b4afc18c4d35/pvs_check.html#fail1

azat commented:
>Stress test (address) — Killed by signal (in clickhouse-server.log)

#28604
